apps/email/js/cards/cmp/autocomplete.js

```diff
@@ -0,0 +1,565 @@
+          var name = contact.name;
+          if (Array.isArray(name)) {
+            name = name[0];
+          }
+
+          // Emails are an array, each one should get its own entry in the
+          // autocomplete.
+          emails.forEach((email) => {
+            var address = email.value;
+
+            // Do not add it if the email address is already an entry.
+            var hasExisting = existingEntries.some(function(entry) {
+              return entry.address === address;
+            });
+            if (hasExisting) {
+              return;
+            }
+
+            var node = itemTemplate.cloneNode(true),
+                detailsNode = node
+                               .querySelector('.cmp-autocomplete-item-details'),
+                nameNode = node.querySelector('.cmp-autocomplete-name'),
+                emailNode = node.querySelector('.cmp-autocomplete-email');
+
+            // Set the ariaLabel for the match, for screen reader use only.
+            mozL10n.setAttributes(detailsNode, 'compose-autocomplete-match',
+                                  { name: name, address: address });
+
+            formatMatchString(queryRegExp,
+                              name || address,
+                              nameNode);
+            formatMatchString(queryRegExp, address|| '', emailNode);
+
+            // Store the match data on the node for ease of access when items
+            // are tapped.
+            var match = {
+              name: name,
+              address: address
+            };
+            node.match = match;
+
+            this.list.appendChild(node);
+          });
+        });
+
+        // Always reshow on a query, since each keystroke has the possibility of
+        // breaking the input field to a new line.
+        if (this.list.children.length) {
+          this.show();
+        } else {
+          this.positionAndHide();
+        }
+      });
+    },
+
+    /**
+     * Positions the input area to the top of the view and sizes the list to
+     * take up as much view space as possible.
+     */
+    positionElements: function() {
+      // Wait to find originTop until there is an actual sizing. Doing this
+      // in the constructor would be incorrect since the initial construction
+      // happens off-screen, so the sizings would incorrectly report zero sizes.
+      if (!this.originTop) {
+        this.originTop = this.originNode.getBoundingClientRect().top;
+      }
+
+      // Want to use the input element, and not a parent of it, for placement of
+      // the autocomplete, since the input element can move to different lines
+      // as typing occurs, where the parent's origin would not change, and would
+      // be a bad reference point to use.
+      var inputRect = this.inputNode.getBoundingClientRect();
+
+      // Scroll the area so that the input item is now at the top of the
+      // address area.
+      var translateY = inputRect.top - this.originTop +
```

Comment:
Given what I saw with the down arrow breakage, maybe it would pay to put scrollTop in here too as defensive thing.  We normally should always be scrollTop=0, but when we're not, it's weird.